New module ConsultDoc

sophiaab opened the pull request:
Functionnality research, browser OK
Billboard : work in progress

jailln commented:
Some GUI related comments first:

* Document Browser:
  * Make it possible to display the browser pane just by checking its check box in the main menu and without having to run a research (to make it possible for users to just navigate through the documents without knowing which documents are stored in the database)
  * When a search is issued in the Research pane:
     * Open Document Browser (if not already) with the search results
     * Add a message in the Document Browser such as `Documents have been updated according to research criteria` 
     * Add a button next to the previous message allowing to reset the research (and thus to navigate in all documents of the database from the Document Browser)
  * Add a close button similarly to the other modules
  * Harmonize the visual appearance of this view (it currently changes when navigating through documents)
  * Add numbers allowing to know how many navigable documents there are and which one we are looking at (e.g. 1/4, 2/4, etc.)

* Research: 
  * Please rename 'Referring date start' to 'Start referring date' (same for end date and publication dates)
  * Add a close button similarly to the other modules

A few other things:
  * What is the purpose of DemoController.html (in comparaison to Demo.html for instance ?) ? Do we need to have another Demo ?
  * Remove billboard related code and GUI elements if not working yet and postpone it to a new pull request (or make it completely work in this PR, depending on the work that still needs to be done on this matter)
  * Remove unused code related to the old (previous to this PR) organization for Documents handling

I will look deeper in the code after these modifications are ok :)

sophiaab commented:
About commit 32a66af :
    Document Browser:
        Close button added 
        Visual appearance harmonized
        Added numbers allowing to know how many navigable documents there are and which one we are looking at (e.g. 1/4, 2/4, etc.)

    Research:
      Renamed dates
      Remove billboard related code and GUI elements

sophiaab commented:
* About commits e3d3406 to 5d856f945
don't pay attention

* About commit 26d7ea7 : 
  * Possible to display the browser pane just by checking its check box in the main menu and without having to run a research 
  * When a search is issued in the Research pane:
Open Document Browser (if not already) with the search results
  * Add close buttons to research window and browser window

jailln commented on UDV-Core/src/Modules/ConsultDoc/DocumentResearch.js:
```diff
@@ -0,0 +1,82 @@
+            ';
+
+        var optionsFilter = "http://rict.liris.cnrs.fr/optionsFilter.json";
+        var schema = "http://rict.liris.cnrs.fr/schemaType.json";
```
The number of configuration files is growing up, I think that it could be a good idea to list them (along with their roles) and to provide examples in a dedicated page pointed in the README.md.

sophiaab replied:
The Readme.md is in UDV-Core/src/Modules/ConsultDoc

jailln commented on UDV-Core/webpack.config.js:
```diff
@@ -22,6 +22,10 @@ module.exports = {
     plugins: [
         definePlugin,
         new webpack.optimize.CommonsChunkPlugin({ name: 'udvcore' }),
+        new webpack.ProvidePlugin({
```
Please add a comment saying that this is related to alpaca

sophiaab replied:
done

jailln commented on UDV-Core/webpack.config.js:
```diff
@@ -62,4 +66,9 @@ module.exports = {
     devServer: {
         publicPath: '/dist/',
     },
+    resolve: {
```
Could you add a comment saying that this is related to alpaca.

sophiaab replied:
done

jailln commented:
General remark:

This is a first step towards a MVC architecture as the code from the views, the controller and the model are separated in different files for the ConsultDoc module. 

However the next step to reach a MVC architecture would be that the views and the model don't know about the controller (which is currently not the case). A way to manage this would be by using events triggered by the views and the model and trapped by the controller. The step after that would be to generalize this approach to the other modules (e.g. temporal) and between modules.

This work is postponed to other pull requests and I will merge this one when the remaining asked modifications are done :)

sophiaab commented on UDV-Core/examples/Demo.js:
```diff
@@ -125,9 +125,31 @@ var help  = new udvcore.HelpWindow({active:true});
-var documents =
-  new udvcore.DocumentsHandler(view,controls,'Vilo3D/docs.csv',{temporal: temporal});
 
+var extendedDocumentModel = {
```
This is planned in a future PR

jailln commented:
Two remarks:
 * If I run a research with Subject:subject 1, the 3 documents are still displayed in the Document Browser when they shoudn't be because from what I can see in the browser they don't have a subject defined. Or maybe they do and then it must be displayed in the Document Browser.
  * Could you limit / harmonize the size of the pictures of documents in the browser ?

EricBoix commented on UDV-Core/examples/DocumentModel.json:
```diff
@@ -0,0 +1,82 @@
+{
```
Would it be possible to regroup DocumentModel.json, optionsFilter.json, and schemaFilters.json within an examples/Document (or Contribute?) sub-directory to indicate they work together ?

sophiaab replied:
For the general demo (with all options / functionnalities), the files are in UDV-Core/examples/data
For the demo regarding consultDoc module, they are in: UDV-Core/src/Modules/ConsultDoc/examples

EricBoix commented on UDV-Core/examples/Demo.js:
```diff
@@ -171,11 +204,21 @@ temporalActiveCtrl.onFinishChange(function(value) {
-documentController.onFinishChange( function(value) { documents.toggleDocBrowserWindow(); });
+//Document uses a folder
+var documentFolder = datDotGUI.addFolder("Documents");
+var docResearch = documentFolder.add( controller.documentResearch, 'windowIsActive').name("Research").listen();
```
Respect the (implicit) coding style and the 80 lines wide constraint.
Manually break (newlines) the lines to make the code more readable !

EricBoix commented on UDV-Core/src/Modules/ConsultDoc/DocumentBrowser.js:
```diff
@@ -0,0 +1,273 @@
+
+import * as THREE from 'three';
```
Same remark as for Demo.js concerning the lines longer than 80 columns.

sophiaab replied:
done

EricBoix replied:
Well, many lines still look much longer than 80 columns to me. Am I wrong ?

sophiaab replied:
Fixed

EricBoix commented on UDV-Core/src/Modules/ConsultDoc/DocumentController.js:
```diff
@@ -0,0 +1,142 @@
+{
+
+    this.url = "http://rict.liris.cnrs.fr/APIVilo3D/APIExtendedDocument/web/"; //url of the server handling documents
+    //FIXME: to put in a configuration file of the general application
```
I just created issue #14 to get this fixed in the next PR.
Issues are much more "sticky" (better reminders) than FIXMEs hidden in the code.

jailln commented on UDV-Core/examples/Demo.js:
```diff
@@ -125,9 +125,42 @@ var help  = new udvcore.HelpWindow({active:true});
 // FIXME For the time being this demo uses the Vilo3D data. Provide a
 // default document for the demo of DocumentHandler class and place it
 // within src/Modules/Documents...
```
test

sophiaab replied:
test

sophiaab replied:
done

sophiaab commented on UDV-Core/examples/Demo.js:
```diff
@@ -171,11 +193,21 @@ temporalActiveCtrl.onFinishChange(function(value) {
-documentController = datDotGUI.add( documents, 'docBrowserWindowIsActive'
-                                    ).name( "Documents" ).listen();
-documentController.onFinishChange( function(value) { documents.toggleDocBrowserWindow(); });
+var billboardOption = documentFolder.add( controller.documentBillboard, 'windowIsActive').name("Billboard(soon)").listen();
```
Done

sophiaab commented on UDV-Core/examples/Demo.html:
```diff
@@ -13,6 +13,7 @@
     <script src="../node_modules/moment/moment.js"></script>
     <script src="../node_modules/three/build/three.js"></script>
     <script src="../src/Externals/dat.gui/dat.gui.min.js"></script>
+    <script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.0/jquery.min.js"></script>
```
Created issue #15

sophiaab commented on UDV-Core/examples/Demo.js:
```diff
@@ -171,11 +204,21 @@ temporalActiveCtrl.onFinishChange(function(value) {
+docResearch.onFinishChange(function(value){
+  controller.documentResearch.refresh();
+});
+/*
```
done

jailln commented:
Good job; we did it! :)
